src/type_checker/checker.rs

```diff
@@ -409,7 +409,7 @@ impl<B: Backend> TypeChecker<B> {
                         .expect("expected a value");
 
                     if let Some(tykind) = &tykind {
-                        if !tykind.same_as(&item_typ.typ) {
+                        if !tykind.match_expected(&item_typ.typ, false) {
```

Comment:
It uses same_as is because it expects the types are not generic and can be matched. The same_as function only consider the case for resolved types.

Also in this case, it doesn't have to be constant to constant match. I think it should allow an array defined as 
```rust
let arr = [var1, 1];
```

So it seems we still need `same_as` here. Or maybe the `match_expected` can have another flag `generic_not_allowed`

Reply:
cc @mimoo